propolis/src/hw/nvme/bits.rs

```diff
@@ -0,0 +1,365 @@
+
+// Version definitions
+
+pub const NVME_VER_1_0: u32 = 0x00010000;
+
+// Admin Command Opcodes
+
+pub const ADMIN_OPC_DELETE_IO_SQ: u8 = 0x00;
+pub const ADMIN_OPC_CREATE_IO_SQ: u8 = 0x01;
+pub const ADMIN_OPC_GET_LOG_PAGE: u8 = 0x02;
+pub const ADMIN_OPC_DELETE_IO_CQ: u8 = 0x04;
+pub const ADMIN_OPC_CREATE_IO_CQ: u8 = 0x05;
+pub const ADMIN_OPC_IDENTIFY: u8 = 0x06;
+pub const ADMIN_OPC_ABORT: u8 = 0x08;
+pub const ADMIN_OPC_SET_FEATURES: u8 = 0x09;
+pub const ADMIN_OPC_GET_FEATURES: u8 = 0x0a;
+pub const ADMIN_OPC_ASYNC_EVENT_REQ: u8 = 0x0c;
+
+// Nvm Command Opcodes
+
+pub const NVM_OPC_FLUSH: u8 = 0x00;
+pub const NVM_OPC_WRITE: u8 = 0x01;
+pub const NVM_OPC_READ: u8 = 0x02;
+
+// Command Status values
+
+pub const STS_SUCCESS: u8 = 0x0;
+pub const STS_INVAL_OPC: u8 = 0x1;
+pub const STS_INVAL_FIELD: u8 = 0x2;
+pub const STS_CID_CONFLICT: u8 = 0x3;
+pub const STS_DATA_XFER_ERR: u8 = 0x4;
+pub const STS_PWR_LOSS_ABRT: u8 = 0x5;
+pub const STS_INTERNAL_ERR: u8 = 0x6;
+pub const STS_ABORT_REQ: u8 = 0x7;
+pub const STS_ABORT_SQ_DEL: u8 = 0x8;
+pub const STS_FAILED_FUSED: u8 = 0x9;
+pub const STS_MISSING_FUSED: u8 = 0xa;
+pub const STS_INVALID_NS: u8 = 0xb;
+pub const STS_COMMAND_SEQ_ERR: u8 = 0xc;
+pub const STS_INVAL_SGL_DESC: u8 = 0xd;
+pub const STS_INVAL_SGL_NUM: u8 = 0xe;
+pub const STS_INVAL_SGL_LEN: u8 = 0xf;
+pub const STS_INVAL_SGL_META_LEN: u8 = 0x10;
+pub const STS_INVAL_SGL_TYPE_LEN: u8 = 0x11;
+pub const STS_INVAL_CMB_USE: u8 = 0x12;
+pub const STS_INVAL_PRP_OFFSET: u8 = 0x13;
+
+pub const STS_CREATE_IO_Q_INVAL_CQ: u8 = 0x0;
+pub const STS_CREATE_IO_Q_INVAL_QID: u8 = 0x1;
+pub const STS_CREATE_IO_Q_INVAL_QSIZE: u8 = 0x2;
+pub const STS_CREATE_IO_Q_INVAL_INT_VEC: u8 = 0x8;
+
+// SetFeature Command Specific Status values
+
+pub const STS_SET_FEATURE_NOT_SAVEABLE: u8 = 0x0D;
+pub const STS_SET_FEATURE_NOT_CHANGEABLE: u8 = 0x0E;
+pub const STS_SET_FEATURE_NOT_NAMESPACE_SPECIFIC: u8 = 0x0F;
+pub const STS_SET_FEATURE_OVERLAPPING_RANGES: u8 = 0x14;
+
+// Read Command Status values
+
+pub const STS_READ_CONFLICTING_ATTRS: u8 = 0x80;
+pub const STS_READ_INVALID_PROT_INFO: u8 = 0x81;
+
+// Feature identifiers
+
+pub const FEAT_ID_ARBITRATION: u8 = 0x01;
+pub const FEAT_ID_POWER_MGMT: u8 = 0x02;
+pub const FEAT_ID_TEMP_THRESH: u8 = 0x04;
+pub const FEAT_ID_ERROR_RECOVERY: u8 = 0x05;
+pub const FEAT_ID_NUM_QUEUES: u8 = 0x07;
+pub const FEAT_ID_INTR_COALESCE: u8 = 0x08;
+pub const FEAT_ID_INTR_VEC_CFG: u8 = 0x09;
+pub const FEAT_ID_WRITE_ATOMIC: u8 = 0x0a;
+pub const FEAT_ID_ASYNC_EVENT_CFG: u8 = 0x0b;
+
+// Identify CNS values
+
+pub const IDENT_CNS_NAMESPACE: u8 = 0x0;
+pub const IDENT_CNS_CONTROLLER: u8 = 0x1;
+pub const IDENT_CNS_ACTIVE_NSID: u8 = 0x2;
+pub const IDENT_CNS_NSID_DESC: u8 = 0x3;
+
+#[derive(Copy, Clone)]
+#[repr(u8)]
+pub enum StatusCodeType {
+    Generic = 0,
+    CmdSpecific = 1,
+    MediaDataIntegrity = 2,
+    VendorSpecific = 7,
+}
+
+#[derive(Default, Copy, Clone)]
+#[repr(C)]
+pub struct PowerStateDescriptor {
```

Comment:
Because other files have a mix of nmve specs that they refer to, I'd suggest we have comments saying what revision of the spec a given table is based on.